### PR TITLE
[state] Collection manipulation

### LINKED ISSUE #326: [state] Add CollectionObservable#mapEach for element-level mapping
After adding [`filter(Function<E, Observable<Boolean>>)`](https://github.com/PoweredByApartium/cocoa-beans/issues/325), a very common next step is to **transform** the elements of a live collection into another type while keeping the same “live view” behavior.

Example: in a game, we often store `GamePlayer` objects, but many systems need `UUID`s, names, or platform `Player` objects, and we want this derived collection to update automatically as the source collection changes.

## Proposed API
```java
public interface CollectionObservable<E, C extends Collection<E>> extends Observable<C> {
    // ...
    <R> CollectionObservable<R, C> mapEach(Function<E, R> mapper);
    // ...
}
```

### LINKED ISSUE #365: [state] Add CollectionObservable#flatMapEach for element-level mapping
After adding [`mapEach(Function<E, R>)`](https://github.com/PoweredByApartium/cocoa-beans/issues/326), a common next step is to transform each element using an observable value while keeping the same “live view” behavior.

`mapEach` is useful when the mapped value can be computed directly from the element.
However, many real use cases need the mapped value to stay updated when some observable state inside the element changes.

Example: in a game, we may store GamePlayer objects, but want a live collection of their current display names, ranks, teams, or platform Player objects. If a player’s name or rank changes, the derived collection should update automatically without removing and re-adding the original GamePlayer.

## Proposed API
```java
public interface CollectionObservable<E, C extends Collection<E>> extends Observable<C> {
         // ...
         <R> CollectionObservable<R, ? extends Collection<R>> flatMapEach(Function<E, Observable<R>> mapper);
         // ...
}
```

### PR DESCRIPTION
**Pull request by ikfir**
### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation)
- [x] 🐞 Bug fix (fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)

### 📚 Description

This PR improves collection manipulation utilities for `CollectionObservable`.

Changes include:

- Ensures `CollectionObservable#filter` returns results in a consistent order.
- Adds `CollectionObservable#mapEach` for live per-element transformations of observable collections.
- Adds `CollectionObservable#flatMapEach` for live per-element mapping to inner observables.

These changes make collection-derived observables more predictable and easier to compose, especially when working with dynamic collections whose elements may update over time.

Closes #326 
Closes #365 

### 🧪 How Has This Been Tested?
Unit testing

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

## Summary by CodeRabbit

* **New Features**
  * Added mapEach and flatMapEach operators for observable collections (lazy, cached per-element transforms and flattened inner-observable mapping); filter now documents a consistent ordering guarantee.

* **Documentation**
  * Added comprehensive guides, examples, and collection-type notes describing mapping, flat-mapping, filtering, subscription semantics, and deduplication behavior.

* **Tests**
  * Large new test suites and samples validating mapEach, flatMapEach, filter behaviors, caching, notification semantics, and edge cases.

### CODE REVIEW

File: cocoa-beans-state/src/main/java/net/apartium/cocoabeans/state/FlatMapElementObservable.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+ * <p>
+ * The class also manages the dynamic updating of its internal collection whenever changes
+ * occur in the source observable or the nested observables it depends upon.
+ *
+ * @param <F> The type of elements in the source observable.
+ * @param <E> The type of elements in the resulting flattened collection.
+ * @param <C> The specific type of collection used to hold the flattened elements.
+ */
+@ApiStatus.AvailableSince("0.0.50")
+public class FlatMapElementObservable<F, E, C extends Collection<E>> implements CollectionObservable<E, C>, Observer {
```
Comment (liorsl) [inline cocoa-beans-state/src/main/java/net/apartium/cocoabeans/state/FlatMapElementObservable.java:22]: you sure this should be an api and not internal?

Comment (ikfir) [reply cocoa-beans-state/src/main/java/net/apartium/cocoabeans/state/FlatMapElementObservable.java:22]: Same as filter

**Review by liorsl — APPROVED** [review]
